tower-balance/src/load/pending_requests.rs

```diff
@@ -0,0 +1,136 @@
+use futures::{Future, Poll, Async};
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use tower::Service;
+use tower_discover::{Change, Discover};
+
+use Load;
+
+/// Expresses `load` based on the number of currently-pending requests.
+#[derive(Debug, Clone)]
+pub struct PendingRequests<T> {
+    inner: T,
+    pending: Arc<AtomicUsize>,
+}
+
+#[derive(Clone, Copy, Debug, Default, PartialOrd, PartialEq, Ord, Eq)]
+pub struct Count(usize);
+
+/// Wraps `inner`'s services with `PendingRequests`.
+pub struct WithPendingRequests<D: Discover> {
+    inner: D,
+}
+
+/// Ensures that `pending` is decremented.
+struct Handle {
+    pending: Arc<AtomicUsize>,
+}
+
+/// Wraps a response future so that it is tracked by `PendingRequests`.
+///
+/// When the inner future completes, either with a value or an error, `pending` is
+/// decremented.
+pub struct ResponseFuture<S: Service> {
+    inner: S::Future,
+    pending: Option<Handle>,
```

Comment:
Why use a Handle with Drop instead of just putting the `Arc<AtomicUsize>` directly on the ResponseFuture struct here?

Reply:
You could do that, but this enforces the desired behavior more correctly (leveraging the type system).

Once `PendingRequests` increments the counter, it produces a Handle.  When the Handle is dropped, no matter how, this ensures that we decrement the counter.  This handles the case where the future is polled as complete and we intentionally drop it.  However, it also handles the case when the Future is dropped before it completes.  We could achieve this by implementing `Drop for ResponseFuture`, but the current implementation is a bit more rigorous imo.